File: Hearthstone-Script/src/main/java/club/xiaojiawei/bean/GameRect.java
package club.xiaojiawei.bean;

import club.xiaojiawei.data.GameRationStaticData;
import club.xiaojiawei.data.ScriptStaticData;
import club.xiaojiawei.utils.GameUtil;
import club.xiaojiawei.utils.RandomUtil;
import club.xiaojiawei.utils.SystemUtil;
import lombok.Getter;
import lombok.ToString;

import java.awt.*;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

/**
 * @author 肖嘉威 xjw580@qq.com
 * @date 2024/8/28 15:15
 */

public record GameRect(double left, double right, double top, double bottom) {

    public static final GameRect INVALID = new GameRect(0, 0, 0, 0);

    public Point getClickPos() {
        int realH = ScriptStaticData.GAME_RECT.bottom - ScriptStaticData.GAME_RECT.top, usableH = realH;
        int realW = ScriptStaticData.GAME_RECT.right - ScriptStaticData.GAME_RECT.left;
        int usableW = (int) (realH * GameRationStaticData.GAME_WINDOW_ASPECT_TO_HEIGHT_RATIO);
        int middleX = realW >> 1;
        int middleY = realH >> 1;
        double pointX = RandomUtil.getRandom(left, right);
        double pointY = RandomUtil.getRandom(top, bottom);
        return new Point((int) (middleX + pointX * usableW), (int) (middleY + pointY * usableH));
    }

    private void cancel(){
        GameUtil.cancelAction();
        SystemUtil.delayTiny();
    }

    public boolean isValid() {
        return !Objects.equals(this, INVALID);
    }

    public void lClick() {
        lClick(true);
    }

    public void lClick(boolean isCancel) {
        if (isCancel) cancel();
        GameUtil.leftButtonClick(getClickPos());
    }

    public void rClick() {
        GameUtil.rightButtonClick(getClickPos());
    }

    public void lClickMoveLClick(GameRect endRect) {
        lClickMoveLClick(endRect, true);
    }

    public void lClickMoveLClick(GameRect endRect, boolean isCancel) {
        if (endRect == null) {
            return;
        }
        if (isCancel) cancel();
        Point startPos = getClickPos();
        Point endPos = endRect.getClickPos();
        GameUtil.leftButtonClick(startPos);
        SystemUtil.delay(100);
        GameUtil.moveMouse(startPos, endPos);
        SystemUtil.delay(100);
        GameUtil.leftButtonClick(endPos);
    }

    public void move() {
        GameUtil.moveMouse(getClickPos());
    }

    public void move(GameRect endRect) {
        if (endRect == null) {
            move();
        } else {
            GameUtil.moveMouse(getClickPos(), endRect.getClickPos());
        }
    }

    public Action buildAction() {
        return new Action(this);
    }

    public static class Action {

        private final GameRect rect;

        private final List<Runnable> runnableList = new ArrayList<>();

        private GameRect lastRect;

        private Action(GameRect rect) {
            this.rect = rect;
        }

        public Action clear() {
            runnableList.clear();
            lastRect = null;
            return this;
        }

        public Action exec() {
            runnableList.forEach(Runnable::run);
            return this;
        }

        public Action lClick() {
            runnableList.add(rect::lClick);
            return this;
        }

        public Action lClick(final GameRect rect) {
            runnableList.add(() -> {
                if (rect == null) {
                    if (lastRect != null) {
                        lastRect.lClick();
                    }
                } else {
                    rect.lClick();
                    lastRect = rect;
                }
            });
            return this;
        }

        public Action rClick() {
            runnableList.add(rect::rClick);
            return this;
        }

        public Action rClick(final GameRect rect) {
            runnableList.add(() -> {
                if (rect == null) {
                    if (lastRect != null) {
                        lastRect.rClick();
                    }
                } else {
                    rect.rClick();
                    lastRect = rect;
                }
            });
            return this;
        }

        public Action move() {
            runnableList.add(rect::move);
            return this;
        }

        public Action move(final GameRect endRect) {
            runnableList.add(() -> rect.move(endRect));
            return this;
        }

    }
}

File: hs-script-app/src/main/java/club/xiaojiawei/hsscript/bean/GameRect.kt
package club.xiaojiawei.hsscript.bean

import club.xiaojiawei.hsscript.consts.GameRationConst
import club.xiaojiawei.hsscript.controller.javafx.GameWindowModalController
import club.xiaojiawei.hsscript.enums.WindowEnum
import club.xiaojiawei.hsscript.status.ScriptStatus
import club.xiaojiawei.hsscript.status.ScriptStatus.GAME_RECT
import club.xiaojiawei.hsscript.utils.GameUtil
import club.xiaojiawei.hsscript.utils.SystemUtil
import club.xiaojiawei.hsscript.utils.WindowUtil
import club.xiaojiawei.hsscript.utils.runUI
import club.xiaojiawei.hsscriptbase.util.RandomUtil
import java.awt.Point
import java.util.function.Consumer

/**
 * @author 肖嘉威 xjw580@qq.com
 * @date 2024/8/28 15:15
 */
data class GameRect(
    val left: Double,
    val right: Double,
    val top: Double,
    val bottom: Double,
) {
    fun getClickPos(): Point {
        val realH: Int = GAME_RECT.bottom - GAME_RECT.top
        val usableH = realH
        val realW: Int = GAME_RECT.right - GAME_RECT.left
        val usableW = (realH * GameRationConst.GAME_WINDOW_ASPECT_TO_HEIGHT_RATIO).toInt()
        val middleX = realW shr 1
        val middleY = realH shr 1
        val pointX: Int = RandomUtil.getRandom((left * usableW).toInt(), (right * usableW).toInt())
        val pointY: Int = RandomUtil.getRandom((top * usableH).toInt(), (bottom * usableH).toInt())
        return Point(middleX + pointX, middleY + pointY)
    }

    private fun cancel() {
        GameUtil.cancelAction()
        SystemUtil.delayTiny()
    }

    fun isValid(): Boolean = this != INVALID

    fun isInvalid(): Boolean = this == INVALID

    private fun showControlPos(gameRect: GameRect = this) {
        if (ScriptStatus.testMode) return
        WindowUtil.getStage(WindowEnum.GAME_WINDOW_CONTROL_MODAL)?.let {
            if (it.isShowing) {
                val controller = WindowUtil.getController(WindowEnum.GAME_WINDOW_CONTROL_MODAL)
                if (controller is GameWindowModalController) {
                    runUI {
                        if (controller.gameRectSize() > 3) {
                            controller.removeFirstGameRect()
                        }
                        controller.drawGameRect(gameRect)
                    }
                }
            }
        }
    }

    /**
     * @param isCancel 执行前先右键取消
     */
    fun lClick(isCancel: Boolean = true) {
        if (isCancel) cancel()
        showControlPos()
        GameUtil.leftButtonClick(getClickPos())
    }

    fun rClick() {
        showControlPos()
        GameUtil.rightButtonClick(getClickPos())
    }

    fun lClickMoveLClick(
        endRect: GameRect?,
        isCancel: Boolean = true,
    ) {
        if (endRect == null) {
            return
        }
        if (isCancel) cancel()
        val startPos = getClickPos()
        val endPos = endRect.getClickPos()
        showControlPos()
        GameUtil.leftButtonClick(startPos)
        SystemUtil.delay(80, 140)
        GameUtil.moveMouse(startPos, endPos)
        SystemUtil.delay(60, 120)
        showControlPos(endRect)
        GameUtil.leftButtonClick(endPos)
    }

    fun move() {
        GameUtil.moveMouse(getClickPos())
    }

    fun move(endRect: GameRect?) {
        if (endRect == null) {
            move()
        } else {
            GameUtil.moveMouse(getClickPos(), endRect.getClickPos())
        }
    }

    fun buildAction(): GameRectAction = GameRectAction(this)

    class GameRectAction(
        private val rect: GameRect,
    ) {
        private val runnableList: MutableList<Runnable?> = mutableListOf()

        private var lastRect: GameRect? = null

        fun clear(): GameRectAction {
            runnableList.clear()
            lastRect = null
            return this
        }

        fun exec(): GameRectAction {
            runnableList.forEach(Consumer { obj: Runnable? -> obj!!.run() })
            return this
        }

        fun lClick(): GameRectAction {
            runnableList.add(Runnable { rect.lClick() })
            return this
        }

        fun lClick(rect: GameRect?): GameRectAction {
            runnableList.add(
                Runnable {
                    if (rect == null) {
                        if (lastRect != null) {
                            lastRect!!.lClick()
                        }
                    } else {
                        rect.lClick()
                        lastRect = rect
                    }
                },
            )
            return this
        }

        fun rClick(): GameRectAction {
            runnableList.add(Runnable { rect.rClick() })
            return this
        }

        fun rClick(rect: GameRect?): GameRectAction {
            runnableList.add(
                Runnable {
                    if (rect == null) {
                        if (lastRect != null) {
                            lastRect!!.rClick()
                        }
                    } else {
                        rect.rClick()
                        lastRect = rect
                    }
                },
            )
            return this
        }

        fun move(): GameRectAction {
            runnableList.add(Runnable { rect.move() })
            return this
        }

        fun move(endRect: GameRect?): GameRectAction {
            runnableList.add(Runnable { rect.move(endRect) })
            return this
        }
    }

    companion object {
        val INVALID: GameRect = GameRect(0.0, 0.0, 0.0, 0.0)
    }
}
